op-supervisor/supervisor/backend/safety/safety.go

```diff
@@ -0,0 +1,274 @@
+	chainID types.ChainID, at eth.L1BlockRef, ref eth.L2BlockRef) error {
+	view, ok := r.safe[chainID]
+	if !ok {
+		iter, err := r.chains.IteratorStartingAt(chainID, ref.Number, 0)
+		if err != nil {
+			return fmt.Errorf("failed to open iterator for chain %s block %d", chainID, ref.Number)
+		}
+		view = &View{
+			chainID: chainID,
+			iter:    iter,
+			localView: heads.HeadPointer{
+				LastSealedBlockHash: ref.Hash,
+				LastSealedBlockNum:  ref.Number,
+				LastSealedTimestamp: ref.Time,
+				LogsSince:           0,
+			},
+			localDerivedFrom: at,
+			validWithinView:  r.ValidWithinSafeView,
+		}
+		r.safe[chainID] = view
+	} else if err := view.UpdateLocal(at, ref); err != nil {
+		return fmt.Errorf("failed to update local-safe: %w", err)
+	}
+
+	// register what this L2 block is derived from
+	m, ok := r.derivedFrom[chainID]
+	if !ok {
+		m = make(map[common.Hash]eth.L1BlockRef)
+		r.derivedFrom[chainID] = m
+	}
+	m[ref.Hash] = at
+	local, _ := r.safe[chainID].Local()
+	r.log.Debug("Updated local safe head", "chainID", chainID, "local", local)
+	r.advanceCrossSafe()
+	return nil
+}
+
+// advanceCrossSafe calls Process on all cross-safe views, and advances the finalized safety status.
+func (r *safetyIndex) advanceCrossSafe() {
+	for chainID, view := range r.safe {
+		if err := view.Process(); err != nil {
+			r.log.Error("Failed to update cross-safe view", "chain", chainID, "err", err)
+		}
+		cross, _ := r.safe[chainID].Cross()
+		r.log.Debug("Updated local safe head", "chainID", chainID, "cross", cross)
+	}
+	r.advanceFinalized()
+	// TODO prune any L2 derivedFrom entry older than the L2 finalized entry
+}
+
+// UpdateFinalizeL1 updates the finalized L1 block, and advances the finalized safety status.
+func (r *safetyIndex) UpdateFinalizeL1(ref eth.L1BlockRef) error {
+	if ref.Number <= r.finalizedL1.Number {
+		return fmt.Errorf("ignoring old L1 finality signal of %s, already have %s", ref, r.finalizedL1)
+	}
+	r.finalizedL1 = ref
+	r.log.Debug("Updated L1 finalized head", "L1finalized", ref)
+	r.advanceFinalized()
+	return nil
+}
+
+// advanceFinalized should be called whenever the finalized L1 block, or the cross-safe history, changes.
+// This then promotes the irreversible cross-safe L2 blocks to a finalized safety status.
+func (r *safetyIndex) advanceFinalized() {
+	// Whatever was considered cross-safe at the finalized block-height can
+	// now be considered finalized, since the inputs have become irreversible.
+	for chainID, view := range r.safe {
+		crossSafe, err := view.Cross()
+		if err != nil {
+			r.log.Info("Failed to get cross-safe data, cannot finalize", "chain", chainID, "err", err)
+			continue
+		}
+		// TODO we need to consider older cross-safe data,
+		//  if we want to finalize something at all on longer lagging finality signal.
+		// Could consider just iterating over all derivedFrom contents?
+		l1Dep := r.derivedFrom[chainID][crossSafe.LastSealedBlockHash]
+		if l1Dep.Number < r.finalizedL1.Number {
```

Comment:
Would be nice to ticket this

Reply:
https://github.com/ethereum-optimism/optimism/issues/12184